npe2/manifest/utils.py

```diff
@@ -153,16 +159,131 @@ def __str__(self) -> str:
         return v
 
 
-def import_python_name(python_name: PythonName) -> Any:
+def _import_npe1_shim(shim_name: str) -> Any:
+    """Import npe1 shimmed python_name
+
+    Some objects returned by npe1 hooks (such as locally defined partials or other
+    objects) don't have globally accessible python names. In such cases, we create
+    a "shim" python_name of the form:
+
+    `__npe1shim__.<hook_python_name>_<index>`
+
+    The implication is that the hook should be imported, called, and indexed to return
+    the corresponding item in the hook results.
+
+    Parameters
+    ----------
+    shim_name : str
+        A string in the form `__npe1shim__.<hook_python_name>_<index>`
+
+    Returns
+    -------
+    Any
+        The <index>th object returned from the callable <hook_python_name>.
+
+    Raises
+    ------
+    IndexError
+        If len(<hook_python_name>()) <= <index>
+    """
+
+    assert shim_name.startswith(SHIM_NAME_PREFIX), f"Invalid shim name: {shim_name}"
+    python_name, idx = shim_name[13:].rsplit("_", maxsplit=1)  # TODO, make a function
+    index = int(idx)
+
+    hook = import_python_name(python_name)
+    result = hook()
+    if isinstance(result, dict):
+        # things like sample_data hookspec return a dict, in which case we want the
+        # "idxth" item in the dict (assumes ordered dict, which is safe now)
+        result = list(result.values())
+    if not isinstance(result, list):
+        result = [result]  # pragma: no cover
+
+    try:
+        out = result[index]
+    except IndexError as e:  # pragma: no cover
+        raise IndexError(f"invalid npe1 shim index {index} for hook {hook}") from e
+
+    if "dock_widget" in python_name and isinstance(out, tuple):
+        return out[0]
+    if "sample_data" in python_name and isinstance(out, dict):
+        # this was a nested sample data
+        return out.get("data")
+
+    return out
+
+
+def import_python_name(python_name: Union[PythonName, str]) -> Any:
     from importlib import import_module
 
-    from ._validators import PYTHON_NAME_PATTERN
+    from . import _validators
 
-    match = PYTHON_NAME_PATTERN.match(python_name)
-    if not match:  # pragma: no cover
-        raise ValueError(f"Invalid python name: {python_name}")
+    if python_name.startswith(SHIM_NAME_PREFIX):
+        return _import_npe1_shim(python_name)
 
-    module_name, funcname = match.groups()
+    _validators.python_name(python_name)  # shows the best error message
+    match = _validators.PYTHON_NAME_PATTERN.match(python_name)
+    module_name, funcname = match.groups()  # type: ignore [union-attr]
```

Comment:
just split on `:` ? I guess using regex to validate even more sure, but if it's invalid it will anyway fail just after.